src/connection.c

```diff
@@ -211,12 +212,71 @@ int update_backoff_delay (backoff_timer_t *timer)
   return timer->delay;
 }  
 
-static void backoff_delay (backoff_timer_t *timer)
-{
-  update_backoff_delay (timer);
-  ParodusInfo("Waiting with backoffRetryTime %d seconds\n", timer->delay);
-  sleep (timer->delay);
-}  
+#define BACKOFF_ERR -2
+#define BACKOFF_SHUTDOWN -1
+#define BACKOFF_IFC_UP   1
+#define BACKOFF_DELAY_TAKEN 0
+
+extern bool interface_down_event;
+void start_conn_in_progress (void);
+
+/* backoff_delay
+ * 
+ * delays for the number of seconds specified in parameter timer
+ * g_shutdown can break out of the delay.
+ * Also, if the interface is down or goes down, then this function
+ * won't return until the interface comes back up, or shutdown is
+ * signalled.  While the interface is down, every 3 minutes the
+ * connection_health_file will be updated.
+ *
+ * returns -2 pthread_cond_timedwait error
+ *  -1   shutdown
+ *  0    delay taken
+ *  1    interface back up
+*/  
+static int backoff_delay (backoff_timer_t *timer)
+{
+  struct timespec ts;
+  int rtn;
+  bool waiting_interface;
+
+  pthread_mutex_lock (get_interface_down_mut());
+  while (true)
+  {
+    clock_gettime (CLOCK_REALTIME, &ts);
+    waiting_interface = interface_down_event;
+    if (!waiting_interface) {
+      update_backoff_delay (timer);
+      ts.tv_sec += timer->delay;
+    } else {
+      init_backoff_timer (timer, -1);
+      ts.tv_sec += 180; /* 3 minutes */
+    }
+    rtn = pthread_cond_timedwait (get_interface_down_con (),
```

Comment:
I feel this timed wait can be added only when waiting_interface is true.

Reply:
As discussed, this seems ok as its same as sleep and in shutdown case, this wait times out as interface down event is reset.